src/common/recordbatch/src/adapter.rs

```diff
@@ -104,3 +106,84 @@ impl Stream for RecordBatchStreamAdapter {
         self.stream.size_hint()
     }
 }
+
+enum AsyncRecordBatchStreamAdapterState {
+    Uninit(
+        Pin<
+            Box<
+                dyn std::future::Future<
+                        Output = std::result::Result<DfSendableRecordBatchStream, DataFusionError>,
+                    > + Send,
+            >,
+        >,
+    ),
+    Inited(std::result::Result<DfSendableRecordBatchStream, DataFusionError>),
+}
+
+pub struct AsyncRecordBatchStreamAdapter {
+    schema: SchemaRef,
+    state: AsyncRecordBatchStreamAdapterState,
+}
+
+impl AsyncRecordBatchStreamAdapter {
+    pub fn new(
+        schema: SchemaRef,
+        stream: Pin<
+            Box<
+                dyn std::future::Future<
+                        Output = std::result::Result<DfSendableRecordBatchStream, DataFusionError>,
+                    > + Send,
+            >,
+        >,
+    ) -> Self {
+        Self {
+            schema,
+            state: AsyncRecordBatchStreamAdapterState::Uninit(stream),
+        }
+    }
+}
+
+impl RecordBatchStream for AsyncRecordBatchStreamAdapter {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+impl Stream for AsyncRecordBatchStreamAdapter {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```

Comment:
We can use [`futures::ready`](https://docs.rs/futures-core/0.3.24/futures_core/macro.ready.html) to simplify delegated future poll like this
```rs
    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        loop {
            match &mut self.state {
                AsyncRecordBatchStreamAdapterState::Uninit(stream_future) => {
                    self.state = AsyncRecordBatchStreamAdapterState::Inited(ready!(Pin::new(
                        stream_future
                    )
                    .poll(cx)));
                    continue;
                }
                AsyncRecordBatchStreamAdapterState::Inited(stream) => match stream {
                    Ok(stream) => {
                        return Poll::Ready(ready!(Pin::new(stream).poll_next(cx)).map(|df| {
                            Ok(RecordBatch {
                                schema: self.schema(),
                                df_recordbatch: df.context(error::PollStreamSnafu)?,
                            })
                        }));
                    }
                    Err(_) => todo!(),
                },
            }
        }
    }
```

Reply:
👍